### DISCUSSION TITLE
Emoji character literals

### DISCUSSION THREAD

**Discussion by ekolis** [dotnet/csharplang#4500] [General]
Using emoji as character literals doesn't always seem to work:

```csharp
public char Glyph { get; } = '🧙🏼‍';
```

This gives a compile error: CS1012 Too many characters in character literal.

However there is only one character (the wizard emoji). I guess certain emoji count as multiple characters? It would be nice if character literals could support those emoji.

**Comment by Phantonia** [comment]
The C# compiler is right. The emoji actually consists of two characters.

The string `"🧙🏼‍"` is internally represented as `"\ud83e\uddd9"`. As you can see, that's two characters. (source: [sharplab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABATARgFgAoE7fABgAIoqBeKgIkD4NwTd3GBuM/ATgAooASi7Fy/KADoAMnAB2AcxgALYUA))

To represent this emoji, you need to use the struct `System.Text.Rune`.

**Comment by jnm2** [comment]
> It would be nice if character literals could support those emoji.

Are you asking for the .NET runtime `char` type to be a different size than 16 bits in order to fit something like `🧙🏼‍` which requires more storage space than 16 bits, or are you asking for the C# compiler to permit character literal syntax such as `'🧙🏼‍'` to be used when assigning to a type other than `char`, such as `System.Text.Rune` which @Phantonia mentioned?

In the meantime, this has the benefit of working today and having possibly even preferable semantics:
```cs
public string Glyph { get; } = "🧙🏼‍";
```

**Reply by ekolis** [reply]
Gee, I suppose either would work. I was thinking of modifying the `char` type to be able to store characters that are bigger than usual, but if there's another type that would work and the syntax could be modified, that might do just as well - but how would `char` be distinguished from `Rune` when creating a literal? Would we need to use an `r` suffix, same as we use `m` with the `decimal` type to distinguish it from `double`? Or maybe backticks could be used in place of quotes?

Yeah, I could use a `string` in place of a `char`, but the context here is a roguelike game - in a traditional roguelike game, everything is represented by an ASCII symbol. I decided to get creative and represent my wizard character with the 🧙🏼‍♂️ emoji. But this wasn't possible with a `char`. If I made it a `string` though - that would allow for something to be represented by multiple characters, so then you might have "hello" on the map or something, occupying multiple tiles, and that would look silly! So I wanted to limit it to a single character. I'll have to look into this `Rune` thing...

**Reply by jnm2** [reply]
@tannergooding comes to mind as a person who could help us understand the magnitude of changing `char` to be larger than 16 bits. I've seen Tanner talking recently about .NET runtime research into keeping strings in memory as UTF8, and maybe there's some crossover here.

> if there's another type that would work and the syntax could be modified, that might do just as well - but how would `char` be distinguished from `Rune` when creating a literal?

It wouldn't need to be distinguished with a suffix. `Rune x = '🧙🏼‍♂️';`  could work in theory, just like how this works:

```cs
FormattableString x = $"This is target-typed as FormattableString instead of String.";
```

**Reply by ufcpp** [reply]
Emojis are unable to be encoded not only in a single `char` (UTF-16 code unit) but also in a single `Rune` (Unicode code point). For example, 👨🏻‍❤️‍💋‍👨🏿 is an emoji represented by 10 code point sequence: 1F468 1F3FB 200D 2764 FE0F 200D 1F48B 200D 1F468 1F3FF which is 10 code points, 15 code units in UTF-16, and 35 bytes in UTF-8. You have to use `System.Globalization.StringInfo` to count emoji correctly.

```cs
using System;
using System.Globalization;
using System.Linq;
using System.Text;

var s = "👨🏻‍❤️‍💋‍👨🏿";
Console.WriteLine(Encoding.UTF8.GetByteCount(s)); // 35
Console.WriteLine(s.Length); // 15
Console.WriteLine(s.EnumerateRunes().Count()); // 10
Console.WriteLine(new StringInfo(s).LengthInTextElements); // 1
```

FYI, there are also some multi-code-point Chinese characters

```cs
using System;
using System.Globalization;
using System.Linq;
using System.Text;

var s = "葛󠄀";
Console.WriteLine(Encoding.UTF8.GetByteCount(s)); // 7
Console.WriteLine(s.Length); // 3
Console.WriteLine(s.EnumerateRunes().Count()); // 2
Console.WriteLine(new StringInfo(s).LengthInTextElements); // 1
```

**Reply by ufcpp** [reply]
Character sequence made from any number of emojis joined by ZWJ is also a single emoji. An example is 👨‍👦‍👦‍👦‍👦‍👦‍👦‍👦‍👦‍👦‍👦 

```cs
using System;
using System.Globalization;
using System.Linq;

const int AnyNumber = 100;
var parent = "👨";
var child = "👦";
var zwj = '\u200D';

var s = parent + zwj + string.Join(zwj, Enumerable.Range(0, AnyNumber).Select(_ => child));

Console.WriteLine(new StringInfo(s).LengthInTextElements); // 1
```

Windows renders this emoji sequence as a single character.

![image](https://user-images.githubusercontent.com/3760955/110226796-54584f80-7f35-11eb-9074-93a98aa0f2f5.png)

**Reply by ufcpp** [reply]
see also: [emoji-zwj-sequences](https://www.unicode.org/Public/emoji/13.1/emoji-zwj-sequences.txt)

🧙🏼‍ has variations for political correctness.

1F9D9 200D 2640 FE0F                        ; RGI_Emoji_ZWJ_Sequence  ; woman mage                                                     # E5.0   [1] (🧙‍♀️)
1F9D9 200D 2642 FE0F                        ; RGI_Emoji_ZWJ_Sequence  ; man mage                                                       # E5.0   [1] (🧙‍♂️)
1F9D9 1F3FB 200D 2640 FE0F                  ; RGI_Emoji_ZWJ_Sequence  ; woman mage: light skin tone                                    # E5.0   [1] (🧙🏻‍♀️)
1F9D9 1F3FB 200D 2642 FE0F                  ; RGI_Emoji_ZWJ_Sequence  ; man mage: light skin tone                                      # E5.0   [1] (🧙🏻‍♂️)
1F9D9 1F3FC 200D 2640 FE0F                  ; RGI_Emoji_ZWJ_Sequence  ; woman mage: medium-light skin tone                             # E5.0   [1] (🧙🏼‍♀️)
1F9D9 1F3FC 200D 2642 FE0F                  ; RGI_Emoji_ZWJ_Sequence  ; man mage: medium-light skin tone                               # E5.0   [1] (🧙🏼‍♂️)
1F9D9 1F3FD 200D 2640 FE0F                  ; RGI_Emoji_ZWJ_Sequence  ; woman mage: medium skin tone                                   # E5.0   [1] (🧙🏽‍♀️)
1F9D9 1F3FD 200D 2642 FE0F                  ; RGI_Emoji_ZWJ_Sequence  ; man mage: medium skin tone                                     # E5.0   [1] (🧙🏽‍♂️)
1F9D9 1F3FE 200D 2640 FE0F                  ; RGI_Emoji_ZWJ_Sequence  ; woman mage: medium-dark skin tone                              # E5.0   [1] (🧙🏾‍♀️)
1F9D9 1F3FE 200D 2642 FE0F                  ; RGI_Emoji_ZWJ_Sequence  ; man mage: medium-dark skin tone                                # E5.0   [1] (🧙🏾‍♂️)
1F9D9 1F3FF 200D 2640 FE0F                  ; RGI_Emoji_ZWJ_Sequence  ; woman mage: dark skin tone                                     # E5.0   [1] (🧙🏿‍♀️)
1F9D9 1F3FF 200D 2642 FE0F                  ; RGI_Emoji_ZWJ_Sequence  ; man mage: dark skin tone                                       # E5.0   [1] (🧙🏿‍♂️)

**Comment by qrli** [comment]
Putting those details aside, I think the core confusion, especially for new generation developers, is the keyword `char`.

As old dogs, we know it originated from C, when English character sets were the only ones used, and `char` made sense back then. Then with more languages supported, we had MBCS (multi-byte character set) so a character might need multiple `char`s. Then initial Unicode tried to use 16 bit `char` to make a character always a `char`, but appeared to fail after Windows, Java, Javascript, etc. were fully on board. And what's worse, a character may even be more than a code point (which may need multiple `char`s to represent), to support combinational characters in some languages. This feature was then reused by emoji to make combined emoji characters, and also abused to form combined non-existing characters, which even poses some security risks.

We are pretty much stuck. If trying to solve it some how, I'd wish a design that the C# language is not coupled to encoding. It would be better to leave those UTF-16/UTF-8 concepts to libraries. 

**Yes, UTF-8 looks like the one true choice today, but so did UTF-16, and so did ASCII.** 🤷
